src/multiclient/promise.h

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include "td/actor/PromiseFuture.h"
+
+namespace multiclient {
+
+template <typename T>
+class PromiseSuccessAny {
+private:
+  struct ControlBlock {
+    ControlBlock(td::Promise<T>&& p) : promise(std::move(p)) {
+    }
+
+    td::Promise<T> promise;
+    int64_t pending_count = 0;
+    std::mutex mutex;
+  };
+
+public:
+  PromiseSuccessAny(td::Promise<T>&& promise) : control_block_(std::make_shared<ControlBlock>(std::move(promise))) {
+  }
+
+  td::Promise<T> get_promise() {
+    control_block_->pending_count += 1;
```

Comment:
it's unlikely to happen but still - in case of some client returns response quickly enough (e.g. cached) that `ctrl->pending_count -= 1;` collide with `control_block_->pending_count += 1;`, we need to wrap it in mutex too